src/sources/internal_logs.rs

```diff
@@ -26,41 +32,95 @@ pub struct InternalLogsConfig {
+    pub pid_key: Option<OptionalValuePath>,
+
+    /// The namespace to use for logs. This overrides the global setting.
+    #[configurable(metadata(docs::hidden))]
+    #[serde(default)]
+    pub log_namespace: Option<bool>,
 }
 
 impl_generate_config_from_default!(InternalLogsConfig);
 
+impl InternalLogsConfig {
+    /// Generates the `schema::Definition` for this component.
+    fn schema_definition(&self, log_namespace: LogNamespace) -> Definition {
+        // `host_key` defaults to the `log_schema().host_key()` if it's not configured in the source.
+        let host_key = self
+            .host_key
+            .clone()
+            .map_or_else(
+                || parse_value_path(log_schema().host_key()).ok(),
+                |k| k.path,
+            )
+            .map(LegacyKey::Overwrite);
+
+        let pid_key = self
+            .pid_key
+            .clone()
+            .and_then(|k| k.path)
+            .map(LegacyKey::Overwrite);
+
+        // There is a global and per-source `log_namespace` config.
+        // The source config overrides the global setting and is merged here.
+        BytesDeserializerConfig
+            .schema_definition(log_namespace)
+            .with_standard_vector_source_metadata()
+            .with_source_metadata(
+                InternalLogsConfig::NAME,
+                host_key,
+                &owned_value_path!("host"),
+                Kind::bytes().or_undefined(),
+                Some("host"),
+            )
+            .with_source_metadata(
+                InternalLogsConfig::NAME,
+                pid_key,
+                &owned_value_path!("pid"),
+                Kind::integer(),
+                Some("pid"),
+            )
+    }
+}
+
 #[async_trait::async_trait]
 impl SourceConfig for InternalLogsConfig {
     async fn build(&self, cx: SourceContext) -> crate::Result<super::Source> {
-        let host_key = self
-            .host_key
-            .as_deref()
-            .unwrap_or_else(|| log_schema().host_key())
-            .to_owned();
-        let pid_key = self.pid_key.as_deref().unwrap_or("pid").to_owned();
+        let host_key = self.host_key.clone().map_or_else(
+            || parse_value_path(log_schema().host_key()).ok(),
+            |k| k.path,
+        );
+
+        let pid_key = self.pid_key.clone().and_then(|k| k.path);
 
         let subscription = TraceSubscription::subscribe();
 
+        let log_namespace = cx.log_namespace(self.log_namespace);
+
         Ok(Box::pin(run(
             host_key,
             pid_key,
             subscription,
             cx.out,
             cx.shutdown,
+            log_namespace,
         )))
     }
 
-    fn outputs(&self, _global_log_namespace: LogNamespace) -> Vec<Output> {
-        vec![Output::default(DataType::Log)]
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<Output> {
+        let schema_definition =
+            self.schema_definition(global_log_namespace.merge(self.log_namespace));
```

Comment:
Do we have a strong preference between having a separate function for `schema_definition()` rather than defining it all just in `outputs()`?

Reply:
yeah good question. I kinda like dropping it in it's own function, but that's just a personal preference thing I guess.
Happy to converge on a design guide for that for consistency though.

Reply:
Yeah, no worries - I've been doing it in the existing function... I think `dnstap` got a separate function, probably a gut check on "size of schema def" 🤷